### PR TITLE
docs: standardized rule docs heading and option-less Options

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

-   [x] Addresses an existing issue: starts on #4365
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Applies two changes to all rules `.md` docs files:

* Changes the top-level heading to be just the rule name, with the description as the following paragraph
* Enforces an `## Options` section with a `jsonc` snippet for just the recommended, non-extension rules without options

<details>
<summary>Fun little script I used to rewrite all rule docs file headings:</summary>

```js
const fs = require('fs');

const dir = fs
  .readdirSync(__dirname)
  .filter(fileName => !fileName.includes('README'));

for (const fileName of dir) {
  const ruleName = fileName.split('.')[0];
  const contents = fs.readFileSync(fileName).toString();
  const firstHeader = contents.match(/^\# (.*)/g)?.[0];
  if (!firstHeader) {
    continue;
  }

  const description = `${firstHeader
    .slice(2)
    .replace(` (\`${ruleName}\`)`, '')}.`;

  const replaced = [
    `# \`${ruleName}\``,
    '',
    description,
    '',
    contents.slice(firstHeader.length + 2),
  ].join('\n');

  fs.writeFileSync(fileName, replaced);
}
```
</details>

### CODE REVIEW

**Comment by JoshuaKGoldberg** [comment]
Drafting pending resolution in https://github.com/typescript-eslint/typescript-eslint/issues/4365.

**Review by bradzacher — COMMENTED** [review]
it'd be good to create a blank `.md` file that people can use as a template when creating a new rule doc. That would make it super easy for people to adhere to the style/layout without going through a test to check.

otherwise people end up copy-pasting an old rule and then sometimes we have bad text in docs.

File: packages/eslint-plugin/docs/rules/TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+const defaultOptions: Options = {
+  someOption: false,
+};
+```
+
+## When Not To Use It
+
+To fill out: why wouldn't you want to use this rule?
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/TEMPLATE.md:50]: "For example if this rule requires a feature released in a certain TS version."

**Review by bradzacher — DISMISSED** [review]

**Review by bradzacher — APPROVED** [review]

**Comment by bradzacher** [comment]
image

**Pull request merged**